fem/particleset.hpp

```diff
@@ -0,0 +1,561 @@
+
+   /// Get const reference to field \p f , component \p c value.
+   const real_t& FieldValue(int f, int c=0) const
+   {
+      MFEM_ASSERT(f >= 0 && static_cast<std::size_t>(f) <  fields.size(),
+                  "invalid field index");
+      MFEM_ASSERT(c >= 0 && c < fields[f].Size(),
+                  "invalid component index");
+      return fields[f][c];
+   }
+
+   /// Get reference to field \p f Vector.
+   Vector& Field(int f)
+   {
+      MFEM_ASSERT(f >= 0 && static_cast<std::size_t>(f) <  fields.size(),
+                  "invalid field index");
+      return fields[f];
+   }
+
+   /// Get const reference to field \p f Vector.
+   const Vector& Field(int f) const
+   {
+      MFEM_ASSERT(f >= 0 && static_cast<std::size_t>(f) <  fields.size(),
+                  "invalid field index");
+      return fields[f];
+   }
+
+   /// Get reference to tag \p t .
+   int& Tag(int t)
+   {
+      MFEM_ASSERT(t >= 0 && static_cast<std::size_t>(t) <  tags.size(),
+                  "invalid tag index");
+      return tags[t][0];
+   }
+
+   /// Get const reference to tag \p t .
+   const int& Tag(int t) const
+   {
+      MFEM_ASSERT(t >= 0 && static_cast<std::size_t>(t) <  tags.size(),
+                  "invalid tag index");
+      return tags[t][0];
+   }
+
+   /// Set tag \p t to reference external data.
+   void SetTagRef(int t, int *tag_data);
+
+   /// Set field \p f to reference external data.
+   void SetFieldRef(int f, real_t *field_data);
+
+   /// Particle equality operator.
+   bool operator==(const Particle &rhs) const;
+
+   /// Particle inequality operator.
+   bool operator!=(const Particle &rhs) const { return !operator==(rhs); }
+
+   /// Print all particle data to \p os.
+   void Print(std::ostream &os=mfem::out) const;
+};
+
+/** @brief ParticleSet initializes and manages data associated with particles.
+ *
+ *  @details Particles are inherently initialized to have a position and an ID,
+ *  and optionally can have any number of Vector (of arbitrary vdim) and scalar
+ *  integer data in the form of @b fields and @b tags respectively. All particle
+ *  data are internally stored in a Struct-of-Arrays fashion, as elaborated on
+ *  below.
+ *
+ *  @par Coordinates:
+ *  All particle coordinates are stored in a ParticleVector with vector
+ *  dimension equal to the spatial dimension, ordered either byNODES or byVDIM.
+ *
+ *  @par IDs:
+ *  Each particle is assigned a unique global ID of type unsigned long long. In
+ *  parallel, IDs are initialized starting with @b rank and striding by @b size.
+ *
+ *  @par Fields:
+ *  Fields represent scalar or vector \ref real_t data to be associated with
+ *  each particles, such as mass, momentum, or moment. For a given field, all
+ *  particle data is stored in a single ParticleVector with a given
+ *  vector dimension (1 for scalar data) and Ordering::Type (byNODES or
+ *  byVDIM).
+ *
+ *  @par Tags:
+ *  Tags represent an integer to be associated with each particle. For a given
```

Comment:
Well `an integer` is actually a vector of Arrays in the `Particle` class. This is confusing.